Failed to launch shop-app: Cannot start meta-application with an abstract environment

I cannot launch the sample app `shop-app`. This exception is thrown:

`Cannot start meta-application with an abstract environment`

**Full log:**

```
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v2.1.5.RELEASE)

2019-05-29 11:18:33.376  INFO 11377 --- [  restartedMain] poussecafe.shop.App                      : Starting App on tswp19 with PID 11377 (/data/bdevos/IdeaProjects/pousse-cafe/pousse-cafe-shop-app/target/classes started by bdevos in /data/bdevos/IdeaProjects/pousse-cafe)
2019-05-29 11:18:33.378 DEBUG 11377 --- [  restartedMain] poussecafe.shop.App                      : Running with Spring Boot v2.1.5.RELEASE, Spring v5.1.7.RELEASE
2019-05-29 11:18:33.378  INFO 11377 --- [  restartedMain] poussecafe.shop.App                      : No active profile set, falling back to default profiles: default
2019-05-29 11:18:33.412  INFO 11377 --- [  restartedMain] .e.DevToolsPropertyDefaultsPostProcessor : Devtools property defaults active! Set 'spring.devtools.add-properties' to 'false' to disable
2019-05-29 11:18:33.412  INFO 11377 --- [  restartedMain] .e.DevToolsPropertyDefaultsPostProcessor : For additional web related logging consider setting the 'logging.level.web' property to 'DEBUG'
2019-05-29 11:18:33.869  INFO 11377 --- [  restartedMain] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data repositories in DEFAULT mode.
2019-05-29 11:18:33.927  INFO 11377 --- [  restartedMain] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 53ms. Found 4 repository interfaces.
2019-05-29 11:18:34.190  INFO 11377 --- [  restartedMain] org.reflections.Reflections              : Reflections took 25 ms to scan 3 urls, producing 48 keys and 89 values 
2019-05-29 11:18:34.211 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public void poussecafe.journal.process.StoreConsumptionResult.storeSuccessfulConsumption(poussecafe.support.model.SuccessfulConsumption)
2019-05-29 11:18:34.211 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public void poussecafe.journal.process.StoreConsumptionResult.storeFailedConsumption(poussecafe.support.model.FailedConsumption)
2019-05-29 11:18:34.213 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.support.adapters.messaging.FailedConsumptionData
2019-05-29 11:18:34.213 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.support.adapters.messaging.SuccessfulConsumptionData
2019-05-29 11:18:34.215 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding data access implementation class poussecafe.journal.adapters.MongoJournalEntryDataAccess
2019-05-29 11:18:34.222  INFO 11377 --- [  restartedMain] org.reflections.Reflections              : Reflections took 4 ms to scan 2 urls, producing 43 keys and 119 values 
2019-05-29 11:18:34.226 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public poussecafe.shop.domain.Customer poussecafe.shop.domain.CustomerFactory.createCustomer(poussecafe.shop.command.CreateCustomer)
2019-05-29 11:18:34.227 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public poussecafe.shop.domain.Product poussecafe.shop.domain.ProductFactory.buildProductWithNoStock(poussecafe.shop.command.CreateProduct)
2019-05-29 11:18:34.227 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public void poussecafe.shop.domain.Order.settle(poussecafe.shop.command.SettleOrder)
2019-05-29 11:18:34.227 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public void poussecafe.shop.domain.Order.ship(poussecafe.shop.command.ShipOrder)
2019-05-29 11:18:34.227 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public void poussecafe.shop.domain.Product.addUnits(poussecafe.shop.command.AddUnits)
2019-05-29 11:18:34.228 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public void poussecafe.shop.domain.Product.placeOrder(poussecafe.shop.command.PlaceOrder)
2019-05-29 11:18:34.228 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public poussecafe.shop.domain.Order poussecafe.shop.domain.OrderFactory.buildPlacedOrder(poussecafe.shop.domain.OrderPlaced)
2019-05-29 11:18:34.228 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public poussecafe.shop.domain.Message poussecafe.shop.domain.MessageFactory.buildMessage(poussecafe.shop.domain.OrderSettled)
2019-05-29 11:18:34.228 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public poussecafe.shop.domain.Message poussecafe.shop.domain.MessageFactory.buildMessage(poussecafe.shop.domain.OrderReadyForShipping)
2019-05-29 11:18:34.228 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public poussecafe.shop.domain.Message poussecafe.shop.domain.MessageFactory.buildMessage(poussecafe.shop.domain.OrderRejected)
2019-05-29 11:18:34.228 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Defining listener for method public poussecafe.shop.domain.Message poussecafe.shop.domain.MessageFactory.buildMessage(poussecafe.shop.domain.OrderCreated)
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.SerializableMessageCreated
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.CreateCustomerData
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.SerializableOrderPlaced
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.SettleOrderData
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.SerializableOrderCreated
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.SerializableOrderReadyForShipping
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.ShipOrderData
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.SerializableOrderRejected
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.AddUnitsData
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.PlaceOrderData
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.SerializableOrderSettled
2019-05-29 11:18:34.230 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding message implementation class poussecafe.shop.adapters.messaging.CreateProductData
2019-05-29 11:18:34.232 DEBUG 11377 --- [  restartedMain] poussecafe.discovery.ClassPathExplorer   : Adding data implementation class poussecafe.shop.domain.mongo.CustomerData
2019-05-29 11:18:34.232 ERROR 11377 --- [  restartedMain] p.environment.EnvironmentBuilder         : 3 abstract entities:
2019-05-29 11:18:34.232 ERROR 11377 --- [  restartedMain] p.environment.EnvironmentBuilder         : - poussecafe.shop.domain.Message
2019-05-29 11:18:34.232 ERROR 11377 --- [  restartedMain] p.environment.EnvironmentBuilder         : - poussecafe.shop.domain.Order
2019-05-29 11:18:34.232 ERROR 11377 --- [  restartedMain] p.environment.EnvironmentBuilder         : - poussecafe.shop.domain.Product
2019-05-29 11:18:34.232 ERROR 11377 --- [  restartedMain] p.environment.EnvironmentBuilder         : 0 abstract messages:
2019-05-29 11:18:34.233 ERROR 11377 --- [  restartedMain] p.environment.EnvironmentBuilder         : 0 abstract services:
2019-05-29 11:18:34.233  WARN 11377 --- [  restartedMain] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'pousseCafeRuntime' defined in class path resource [poussecafe/shop/AppConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [poussecafe.runtime.Runtime]: Factory method 'pousseCafeRuntime' threw exception; nested exception is poussecafe.exception.PousseCafeException: Cannot start meta-application with an abstract environment
2019-05-29 11:18:34.238  INFO 11377 --- [  restartedMain] ConditionEvaluationReportLoggingListener : 

Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2019-05-29 11:18:34.243 ERROR 11377 --- [  restartedMain] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'pousseCafeRuntime' defined in class path resource [poussecafe/shop/AppConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [poussecafe.runtime.Runtime]: Factory method 'pousseCafeRuntime' threw exception; nested exception is poussecafe.exception.PousseCafeException: Cannot start meta-application with an abstract environment
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:627) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:456) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1321) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1160) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:224) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveNamedBean(DefaultListableBeanFactory.java:1116) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveBean(DefaultListableBeanFactory.java:409) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:343) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:337) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at poussecafe.spring.SpringBridge.postProcessBeanFactory(SpringBridge.java:16) ~[classes/:na]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:286) ~[spring-context-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:181) ~[spring-context-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:705) ~[spring-context-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:531) ~[spring-context-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:142) ~[spring-boot-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775) ~[spring-boot-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397) ~[spring-boot-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316) ~[spring-boot-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1260) ~[spring-boot-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1248) ~[spring-boot-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at poussecafe.shop.App.main(App.java:12) ~[classes/:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49) ~[spring-boot-devtools-2.1.5.RELEASE.jar:2.1.5.RELEASE]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [poussecafe.runtime.Runtime]: Factory method 'pousseCafeRuntime' threw exception; nested exception is poussecafe.exception.PousseCafeException: Cannot start meta-application with an abstract environment
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:622) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	... 30 common frames omitted
Caused by: poussecafe.exception.PousseCafeException: Cannot start meta-application with an abstract environment
	at poussecafe.environment.EnvironmentBuilder.checkEnvironment(EnvironmentBuilder.java:189) ~[classes/:na]
	at poussecafe.environment.EnvironmentBuilder.build(EnvironmentBuilder.java:156) ~[classes/:na]
	at poussecafe.runtime.Runtime$Builder.build(Runtime.java:82) ~[classes/:na]
	at poussecafe.shop.AppConfiguration.pousseCafeRuntime(AppConfiguration.java:31) ~[classes/:na]
	at poussecafe.shop.AppConfiguration$$EnhancerBySpringCGLIB$$b12c4d6e.CGLIB$pousseCafeRuntime$0(<generated>) ~[classes/:na]
	at poussecafe.shop.AppConfiguration$$EnhancerBySpringCGLIB$$b12c4d6e$$FastClassBySpringCGLIB$$554b83ed.invoke(<generated>) ~[classes/:na]
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244) ~[spring-core-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:363) ~[spring-context-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	at poussecafe.shop.AppConfiguration$$EnhancerBySpringCGLIB$$b12c4d6e.pousseCafeRuntime(<generated>) ~[classes/:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.1.7.RELEASE.jar:5.1.7.RELEASE]
	... 31 common frames omitted```

@gdethier The 2 commits (https://github.com/benoitdevos/pousse-cafe/commit/580f6748dda2430cb5db0f60c527b561fd3edac2 and https://github.com/benoitdevos/pousse-cafe/commit/1d395534ac9a01bdde02c7076ec51c33195d4d98) allow the shop app to start. Then it's possible to POST a new product and GET it as well.

One problem remains: When a product is created, the log is filled with these lines:

```2019-06-04 20:51:28.786 DEBUG 20898 --- [l message queue] poussecafe.runtime.MessageConsumer       : Handling received message poussecafe.support.adapters.messaging.SuccessfulConsumptionData@ef90b73
2019-06-04 20:51:28.786 DEBUG 20898 --- [l message queue] poussecafe.runtime.MessageConsumer       : Found 1 listeners
2019-06-04 20:51:28.786 DEBUG 20898 --- [l message queue] poussecafe.runtime.MessageConsumer       : Consumption of message poussecafe.support.adapters.messaging.SuccessfulConsumptionData@ef90b73 by listener poussecafe.environment.MessageListener@583a4ec0[poussecafe.journal.process.StoreConsumptionResult::storeSuccessfulConsumption(poussecafe.support.model.SuccessfulConsumption)]
2019-06-04 20:51:28.786 DEBUG 20898 --- [l message queue] poussecafe.runtime.MessageConsumer       : Consumption of message poussecafe.support.adapters.messaging.SuccessfulConsumptionData@ef90b73 by listener poussecafe.environment.MessageListener@583a4ec0[poussecafe.journal.process.StoreConsumptionResult::storeSuccessfulConsumption(poussecafe.support.model.SuccessfulConsumption)] succeeded
2019-06-04 20:51:28.786 DEBUG 20898 --- [l message queue] poussecafe.runtime.MessageConsumer       : Message poussecafe.support.adapters.messaging.SuccessfulConsumptionData@ef90b73 handled (consumption ID 2e4974ba-9e36-43c5-839f-2aecb4db4d48)
```

Then when the app is stopped, the following lines appear in the log;
```
2019-06-04 20:51:29.203 DEBUG 20898 --- [l message queue] poussecafe.runtime.MessageConsumer       : Message poussecafe.support.adapters.messaging.SuccessfulConsumptionData@44fad4cd handled (consumption ID b34fa983-a6e3-4a7c-8d08-22abb6a369dc)
2019-06-04 20:51:29.204 DEBUG 20898 --- [l message queue] poussecafe.runtime.MessageConsumer       : Handling received message poussecafe.support.adapters.messaging.FailedConsumptionData@7923446d
2019-06-04 20:51:29.204 DEBUG 20898 --- [l message queue] poussecafe.runtime.MessageConsumer       : Found 1 listeners
2019-06-04 20:51:29.204 DEBUG 20898 --- [l message queue] poussecafe.runtime.MessageConsumer       : Consumption of message poussecafe.support.adapters.messaging.FailedConsumptionData@7923446d by listener poussecafe.environment.MessageListener@4e5e8ba6[poussecafe.journal.process.StoreConsumptionResult::storeFailedConsumption(poussecafe.support.model.FailedConsumption)]
2019-06-04 20:51:29.212 ERROR 20898 --- [l message queue] poussecafe.runtime.MessageConsumer       : Consumption of message poussecafe.support.adapters.messaging.FailedConsumptionData@7923446d by listener poussecafe.environment.MessageListener@4e5e8ba6[poussecafe.journal.process.StoreConsumptionResult::storeFailedConsumption(poussecafe.support.model.FailedConsumption)] failed

poussecafe.exception.PousseCafeException: Unable to invoke method
    at poussecafe.util.MethodInvoker.invoke(MethodInvoker.java:38) ~[classes/:na]
    at poussecafe.environment.DomainProcessMessageListenerFactory.lambda$buildMessageListener$0(DomainProcessMessageListenerFactory.java:24) ~[classes/:na]
    at poussecafe.runtime.MessageConsumer.consumeMessage(MessageConsumer.java:117) ~[classes/:na]
    at poussecafe.runtime.MessageConsumer.consumeMessage(MessageConsumer.java:85) ~[classes/:na]
    at poussecafe.runtime.MessageConsumer.consumeMessage(MessageConsumer.java:66) ~[classes/:na]
    at poussecafe.messaging.MessageReceiver.onMessage(MessageReceiver.java:29) ~[classes/:na]
    at poussecafe.messaging.internal.InternalMessagingQueue$InternalMessageReceiver.lambda$actuallyStartReceiving$0(InternalMessagingQueue.java:38) ~[classes/:na]
    at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
Caused by: java.lang.reflect.InvocationTargetException: null
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
    at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
    at poussecafe.util.MethodInvoker.invoke(MethodInvoker.java:36) ~[classes/:na]
    ... 7 common frames omitted
Caused by: java.lang.IllegalStateException: state should be: open
    at com.mongodb.assertions.Assertions.isTrue(Assertions.java:70) ~[mongodb-driver-core-3.8.2.jar:na]
    at com.mongodb.internal.connection.BaseCluster.getDescription(BaseCluster.java:164) ~[mongodb-driver-core-3.8.2.jar:na]
    at com.mongodb.internal.connection.SingleServerCluster.getDescription(SingleServerCluster.java:41) ~[mongodb-driver-core-3.8.2.jar:na]
    at com.mongodb.client.internal.MongoClientDelegate.getConnectedClusterDescription(MongoClientDelegate.java:136) ~[mongodb-driver-3.8.2.jar:na]
    at com.mongodb.client.internal.MongoClientDelegate.createClientSession(MongoClientDelegate.java:94) ~[mongodb-driver-3.8.2.jar:na]
    at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.getClientSession(MongoClientDelegate.java:249) ~[mongodb-driver-3.8.2.jar:na]
    at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:190) ~[mongodb-driver-3.8.2.jar:na]
    at com.mongodb.client.internal.MongoCollectionImpl.executeSingleWriteRequest(MongoCollectionImpl.java:960) ~[mongodb-driver-3.8.2.jar:na]
    at com.mongodb.client.internal.MongoCollectionImpl.executeInsertOne(MongoCollectionImpl.java:494) ~[mongodb-driver-3.8.2.jar:na]
    at com.mongodb.client.internal.MongoCollectionImpl.insertOne(MongoCollectionImpl.java:478) ~[mongodb-driver-3.8.2.jar:na]
    at com.mongodb.client.internal.MongoCollectionImpl.insertOne(MongoCollectionImpl.java:472) ~[mongodb-driver-3.8.2.jar:na]
    at org.springframework.data.mongodb.core.MongoTemplate$6.doInCollection(MongoTemplate.java:1443) ~[spring-data-mongodb-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.data.mongodb.core.MongoTemplate.execute(MongoTemplate.java:545) ~[spring-data-mongodb-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.data.mongodb.core.MongoTemplate.insertDocument(MongoTemplate.java:1436) ~[spring-data-mongodb-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.data.mongodb.core.MongoTemplate.doInsert(MongoTemplate.java:1244) ~[spring-data-mongodb-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.data.mongodb.core.MongoTemplate.insert(MongoTemplate.java:1178) ~[spring-data-mongodb-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.data.mongodb.repository.support.SimpleMongoRepository.insert(SimpleMongoRepository.java:244) ~[spring-data-mongodb-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at jdk.internal.reflect.GeneratedMethodAccessor37.invoke(Unknown Source) ~[na:na]
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
    at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
    at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:359) ~[spring-data-commons-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:200) ~[spring-data-commons-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:644) ~[spring-data-commons-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.7.RELEASE.jar:5.1.7.RELEASE]
    at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:608) ~[spring-data-commons-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.lambda$invoke$3(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.7.RELEASE.jar:5.1.7.RELEASE]
    at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:59) ~[spring-data-commons-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.7.RELEASE.jar:5.1.7.RELEASE]
    at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93) ~[spring-aop-5.1.7.RELEASE.jar:5.1.7.RELEASE]
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.7.RELEASE.jar:5.1.7.RELEASE]
    at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:61) ~[spring-data-commons-2.1.8.RELEASE.jar:2.1.8.RELEASE]
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.7.RELEASE.jar:5.1.7.RELEASE]
    at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212) ~[spring-aop-5.1.7.RELEASE.jar:5.1.7.RELEASE]
    at com.sun.proxy.$Proxy72.insert(Unknown Source) ~[na:na]
    at poussecafe.spring.mongo.storage.MongoDataAccess.addData(MongoDataAccess.java:25) ~[classes/:na]
    at poussecafe.domain.Repository.addData(Repository.java:70) ~[classes/:na]
    at poussecafe.domain.Repository.add(Repository.java:63) ~[classes/:na]
    at poussecafe.journal.process.StoreConsumptionResult.lambda$storeFailedConsumption$1(StoreConsumptionResult.java:26) ~[classes/:na]
    at poussecafe.storage.NoTransactionRunner.runInTransaction(NoTransactionRunner.java:7) ~[classes/:na]
    at poussecafe.process.TransactionAwareService.runInTransaction(TransactionAwareService.java:11) ~[classes/:na]
    at poussecafe.journal.process.StoreConsumptionResult.storeFailedConsumption(StoreConsumptionResult.java:26) ~[classes/:na]
    ... 12 common frames omitted
```

--- a/pousse-cafe-journal-spring-mongo/src/main/java/poussecafe/journal/adapters/MongoJournalEntryDataAccess.java
+++ b/pousse-cafe-journal-spring-mongo/src/main/java/poussecafe/journal/adapters/MongoJournalEntryDataAccess.java
@@ -4,10 +4,9 @@ import java.util.List;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.data.mongodb.repository.MongoRepository;
 import poussecafe.discovery.DataAccessImplementation;
-import poussecafe.journal.adapters.SerializableJournalEntryId;
+import poussecafe.journal.domain.ConsumptionStatus;
 import poussecafe.journal.domain.JournalEntry;
 import poussecafe.journal.domain.JournalEntryId;
-import poussecafe.journal.domain.ConsumptionStatus;
 import poussecafe.spring.mongo.storage.MongoDataAccess;
 import poussecafe.spring.mongo.storage.SpringMongoDbStorage;
 
@@ -23,18 +22,13 @@ public class MongoJournalEntryDataAccess extends MongoDataAccess<JournalEntryId,
     private MongoJournalEntryDataMongoRepository repository;
 
     @Override
-    public List<MongoJournalEntryData> findByMessageId(String messageId) {
-        return repository.findByMessageId(messageId);
-    }
-
-    @Override
     public List<MongoJournalEntryData> findByStatus(ConsumptionStatus status) {
         return repository.findByStatus(status);
     }
 
     @Override
     protected SerializableJournalEntryId convertId(JournalEntryId id) {
-        return new SerializableJournalEntryId(id);
+        return SerializableJournalEntryId.adapt(id);
     }
 
     @Override

--- a/pousse-cafe-journal-spring-mongo/src/main/java/poussecafe/journal/adapters/MongoJournalEntryDataMongoRepository.java
+++ b/pousse-cafe-journal-spring-mongo/src/main/java/poussecafe/journal/adapters/MongoJournalEntryDataMongoRepository.java
@@ -2,12 +2,9 @@ package poussecafe.journal.adapters;
 
 import java.util.List;
 import org.springframework.data.mongodb.repository.MongoRepository;
-import poussecafe.journal.adapters.SerializableJournalEntryId;
 import poussecafe.journal.domain.ConsumptionStatus;
 
 public interface MongoJournalEntryDataMongoRepository extends MongoRepository<MongoJournalEntryData, SerializableJournalEntryId> {
-
-    List<MongoJournalEntryData> findByMessageId(String messageId);
 
     List<MongoJournalEntryData> findByStatus(ConsumptionStatus status);
 

--- a/pousse-cafe-journal/src/main/java/poussecafe/journal/adapters/InternalJournalEntryDataAccess.java
+++ b/pousse-cafe-journal/src/main/java/poussecafe/journal/adapters/InternalJournalEntryDataAccess.java
@@ -23,11 +23,6 @@ public class InternalJournalEntryDataAccess extends InternalDataAccess<JournalEn
     }
 
     @Override
-    public List<JournalEntryData> findByMessageId(String messageId) {
-        return findBy(messageId);
-    }
-
-    @Override
     public List<JournalEntryData> findByStatus(ConsumptionStatus status) {
         return findBy(status);
     }

--- a/pousse-cafe-journal/src/main/java/poussecafe/journal/adapters/SerializableJournalEntryId.java
+++ b/pousse-cafe-journal/src/main/java/poussecafe/journal/adapters/SerializableJournalEntryId.java
@@ -8,17 +8,17 @@ public class SerializableJournalEntryId implements Serializable {
 
     public static SerializableJournalEntryId adapt(JournalEntryId to) {
         SerializableJournalEntryId data = new SerializableJournalEntryId();
-        data.messageId = to.getConsumptionId();
+        data.consumptionId = to.getConsumptionId();
         data.listenerId = to.getListenerId();
         return data;
     }
 
-    private String messageId;
+    private String consumptionId;
 
     private String listenerId;
 
     public JournalEntryId adapt() {
-        return new JournalEntryId(messageId, listenerId);
+        return new JournalEntryId(consumptionId, listenerId);
     }
 
 }

--- a/pousse-cafe-journal/src/main/java/poussecafe/journal/adapters/SerializableJournalEntryId.java
+++ b/pousse-cafe-journal/src/main/java/poussecafe/journal/adapters/SerializableJournalEntryId.java
@@ -6,16 +6,18 @@ import poussecafe.journal.domain.JournalEntryId;
 @SuppressWarnings("serial")
 public class SerializableJournalEntryId implements Serializable {
 
-    public SerializableJournalEntryId(JournalEntryId to) {
-        messageId = to.getConsumptionId();
-        listenerId = to.getListenerId();
+    public static SerializableJournalEntryId adapt(JournalEntryId to) {
+        SerializableJournalEntryId data = new SerializableJournalEntryId();
+        data.messageId = to.getConsumptionId();
+        data.listenerId = to.getListenerId();
+        return data;
     }
 
     private String messageId;
 
     private String listenerId;
 
-    public JournalEntryId toJournalEntryId() {
+    public JournalEntryId adapt() {
         return new JournalEntryId(messageId, listenerId);
     }
 

--- a/pousse-cafe-journal/src/main/java/poussecafe/journal/domain/JournalEntryDataAccess.java
+++ b/pousse-cafe-journal/src/main/java/poussecafe/journal/domain/JournalEntryDataAccess.java
@@ -5,7 +5,5 @@ import poussecafe.domain.EntityDataAccess;
 
 public interface JournalEntryDataAccess<D extends JournalEntry.Attributes> extends EntityDataAccess<JournalEntryId, D> {
 
-    List<D> findByMessageId(String messageId);
-
     List<D> findByStatus(ConsumptionStatus status);
 }

--- a/pousse-cafe-journal/src/main/java/poussecafe/journal/domain/JournalEntryRepository.java
+++ b/pousse-cafe-journal/src/main/java/poussecafe/journal/domain/JournalEntryRepository.java
@@ -5,10 +5,6 @@ import poussecafe.domain.Repository;
 
 public class JournalEntryRepository extends Repository<JournalEntry, JournalEntryId, JournalEntry.Attributes> {
 
-    public List<JournalEntry> findByConsumptionId(String messageId) {
-        return wrap(dataAccess().findByMessageId(messageId));
-    }
-
     @Override
     public JournalEntryDataAccess<JournalEntry.Attributes> dataAccess() {
         return (JournalEntryDataAccess<JournalEntry.Attributes>) super.dataAccess();

--- a/pousse-cafe-shop-app/README.md
+++ b/pousse-cafe-shop-app/README.md
@@ -15,8 +15,8 @@ for the common application properties).
 1. Build and launch the app
 2. Create a product with the following command:
 
-    curl -X POST "http://localhost:8080/product" -d '{"key":"id"}' -H "Content-Type: application/json"
+    curl -X POST "http://localhost:8080/product" -d '{"id":"product-123"}' -H "Content-Type: application/json"
 
 3. Fetch the product:
 
-    curl -X GET "http://localhost:8080/product/id"
+    curl -X GET "http://localhost:8080/product/product-123"

--- a/pousse-cafe-shop-app/src/main/java/poussecafe/shop/App.java
+++ b/pousse-cafe-shop-app/src/main/java/poussecafe/shop/App.java
@@ -5,7 +5,7 @@ import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.data.mongodb.repository.config.EnableMongoRepositories;
 
 @SpringBootApplication
-@EnableMongoRepositories(basePackages = {"poussecafe.shop", "poussecafe.spring.mongo"})
+@EnableMongoRepositories(basePackages = {"poussecafe.shop", "poussecafe.spring.mongo", "poussecafe.journal"})
 public class App {
 
     public static void main(String[] args) {

--- a/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/CustomerData.java
+++ b/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/CustomerData.java
@@ -2,15 +2,9 @@ package poussecafe.shop.domain.mongo;
 
 import org.springframework.data.annotation.Id;
 import poussecafe.attribute.Attribute;
-import poussecafe.discovery.DataImplementation;
 import poussecafe.shop.domain.Customer;
 import poussecafe.shop.domain.CustomerId;
-import poussecafe.spring.mongo.storage.SpringMongoDbStorage;
 
-@DataImplementation(
-    entity = Customer.class,
-    storageNames = SpringMongoDbStorage.NAME
-)
 public class CustomerData implements Customer.Attributes {
 
     @Override

--- a/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/CustomerDataAccess.java
+++ b/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/CustomerDataAccess.java
@@ -2,10 +2,17 @@ package poussecafe.shop.domain.mongo;
 
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.data.mongodb.repository.MongoRepository;
+import poussecafe.discovery.DataAccessImplementation;
+import poussecafe.shop.domain.Customer;
 import poussecafe.shop.domain.CustomerId;
 import poussecafe.spring.mongo.storage.MongoDataAccess;
+import poussecafe.spring.mongo.storage.SpringMongoDbStorage;
 
-public class CustomerDataAccess extends MongoDataAccess<CustomerId, CustomerData, String> {
+@DataAccessImplementation(
+        aggregateRoot = Customer.class,
+        dataImplementation = CustomerData.class,
+        storageName = SpringMongoDbStorage.NAME
+)public class CustomerDataAccess extends MongoDataAccess<CustomerId, CustomerData, String> implements poussecafe.shop.domain.CustomerDataAccess<CustomerData> {
 
     @Autowired
     private CustomerMongoRepository repository;

--- a/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/CustomerDataAccess.java
+++ b/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/CustomerDataAccess.java
@@ -12,7 +12,8 @@ import poussecafe.spring.mongo.storage.SpringMongoDbStorage;
         aggregateRoot = Customer.class,
         dataImplementation = CustomerData.class,
         storageName = SpringMongoDbStorage.NAME
-)public class CustomerDataAccess extends MongoDataAccess<CustomerId, CustomerData, String> implements poussecafe.shop.domain.CustomerDataAccess<CustomerData> {
+)
+public class CustomerDataAccess extends MongoDataAccess<CustomerId, CustomerData, String> implements poussecafe.shop.domain.CustomerDataAccess<CustomerData> {
 
     @Autowired
     private CustomerMongoRepository repository;

--- a/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/MessageDataAccess.java
+++ b/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/MessageDataAccess.java
@@ -3,10 +3,18 @@ package poussecafe.shop.domain.mongo;
 import java.util.List;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.data.mongodb.repository.MongoRepository;
+import poussecafe.discovery.DataAccessImplementation;
 import poussecafe.shop.domain.CustomerId;
+import poussecafe.shop.domain.Message;
 import poussecafe.shop.domain.MessageId;
 import poussecafe.spring.mongo.storage.MongoDataAccess;
+import poussecafe.spring.mongo.storage.SpringMongoDbStorage;
 
+@DataAccessImplementation(
+        aggregateRoot = Message.class,
+        dataImplementation = MessageData.class,
+        storageName = SpringMongoDbStorage.NAME
+)
 public class MessageDataAccess extends MongoDataAccess<MessageId, MessageData, String> implements poussecafe.shop.domain.MessageDataAccess<MessageData> {
 
     @Autowired

--- a/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/OrderDataAccess.java
+++ b/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/OrderDataAccess.java
@@ -2,11 +2,19 @@ package poussecafe.shop.domain.mongo;
 
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.data.mongodb.repository.MongoRepository;
+import poussecafe.discovery.DataAccessImplementation;
 import poussecafe.shop.adapters.storage.OrderIdData;
+import poussecafe.shop.domain.Order;
 import poussecafe.shop.domain.OrderId;
 import poussecafe.spring.mongo.storage.MongoDataAccess;
+import poussecafe.spring.mongo.storage.SpringMongoDbStorage;
 
-public class OrderDataAccess extends MongoDataAccess<OrderId, OrderData, OrderIdData> {
+@DataAccessImplementation(
+        aggregateRoot = Order.class,
+        dataImplementation = OrderData.class,
+        storageName = SpringMongoDbStorage.NAME
+)
+public class OrderDataAccess extends MongoDataAccess<OrderId, OrderData, OrderIdData> implements poussecafe.shop.domain.OrderDataAccess<OrderData> {
 
     @Autowired
     private OrderMongoRepository repository;

--- a/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/ProductDataAccess.java
+++ b/pousse-cafe-shop-app/src/main/java/poussecafe/shop/domain/mongo/ProductDataAccess.java
@@ -2,10 +2,18 @@ package poussecafe.shop.domain.mongo;
 
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.data.mongodb.repository.MongoRepository;
+import poussecafe.discovery.DataAccessImplementation;
+import poussecafe.shop.domain.Product;
 import poussecafe.shop.domain.ProductId;
 import poussecafe.spring.mongo.storage.MongoDataAccess;
+import poussecafe.spring.mongo.storage.SpringMongoDbStorage;
 
-public class ProductDataAccess extends MongoDataAccess<ProductId, ProductData, String> {
+@DataAccessImplementation(
+        aggregateRoot = Product.class,
+        dataImplementation = ProductData.class,
+        storageName = SpringMongoDbStorage.NAME
+)
+public class ProductDataAccess extends MongoDataAccess<ProductId, ProductData, String> implements poussecafe.shop.domain.ProductDataAccess<ProductData> {
 
     @Autowired
     private ProductMongoRepository repository;

--- a/pousse-cafe-shop/src/main/java/poussecafe/shop/domain/CustomerDataAccess.java
+++ b/pousse-cafe-shop/src/main/java/poussecafe/shop/domain/CustomerDataAccess.java
@@ -1,0 +1,6 @@
+package poussecafe.shop.domain;
+
+import poussecafe.domain.EntityDataAccess;
+
+public interface CustomerDataAccess<N extends Customer.Attributes> extends EntityDataAccess<CustomerId, N> {
+}

--- a/pousse-cafe-shop/src/main/java/poussecafe/shop/domain/OrderDataAccess.java
+++ b/pousse-cafe-shop/src/main/java/poussecafe/shop/domain/OrderDataAccess.java
@@ -1,0 +1,6 @@
+package poussecafe.shop.domain;
+
+import poussecafe.domain.EntityDataAccess;
+
+public interface OrderDataAccess<N extends Order.Attributes> extends EntityDataAccess<OrderId, N> {
+}

--- a/pousse-cafe-shop/src/main/java/poussecafe/shop/domain/ProductDataAccess.java
+++ b/pousse-cafe-shop/src/main/java/poussecafe/shop/domain/ProductDataAccess.java
@@ -1,0 +1,6 @@
+package poussecafe.shop.domain;
+
+import poussecafe.domain.EntityDataAccess;
+
+public interface ProductDataAccess<N extends Product.Attributes> extends EntityDataAccess<ProductId, N> {
+}

